### PR TITLE
Add FloodForecaster: Domain-Adaptive GINO Framework for Flood Forecasting

### PR DESCRIPTION
**Pull request by MehdiTaghizadehUVa**
# PhysicsNeMo Pull Request

## Description

This PR integrates **FloodForecaster**, a domain-adaptive Geometry-Informed Neural Operator (GINO) framework for rapid, high-resolution flood forecasting. The framework enables accurate, real-time flood predictions by learning from source domain data and adapting to target domains through adversarial training.

### Key Features

- **Domain-Adaptive Training**: Three-stage pipeline (pretraining → domain adaptation → rollout evaluation) for transfer learning from data-rich to data-scarce domains
- **Gradient Reversal Layer (GRL)**: Adversarial domain adaptation using CNN-based domain classifier
- **GINO Integration**: Combines Graph Neural Operators (GNO) and Fourier Neural Operators (FNO) for irregular terrain processing
- **Physics-Informed Metrics**: Volume conservation, arrival time, inundation duration, CSI, and FHCA
- **PhysicsNeMo Module Compliance**: All components inherit from `physicsnemo.Module` with checkpointing support

### Components Added

- Training modules: domain adaptation trainer, pretraining pipeline
- Data processing: GINO wrapper, preprocessing/postprocessing utilities
- Datasets: Custom dataset classes for flood data loading
- Inference: Rollout evaluation and visualization pipeline
- Configuration: Hydra-based configuration system
- Documentation: Complete README with usage examples

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Dependencies

No new dependencies required. All packages are either already in PhysicsNeMo or standard scientific computing libraries:
- `neuralop>=2.0.0` (existing)
- `hydra-core>=1.2.0` (existing)
- `wandb>=0.12.0` (optional, for logging)
- Standard packages: `matplotlib`, `tqdm`, `numpy`, `torch`, `omegaconf`, `pandas`, `h5py`

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: untrained_checkpoint.mdlus
Comment (mnabian) [inline untrained_checkpoint.mdlus:1]: Please delete this file

File: physicsnemo/models/module.py [hunk 1]
```diff
@@ -709,7 +709,7 @@ def load(
                 model_bytes = archive.read("model.pt")
 
             # Load state dict after closing archive
-            model_dict = torch.load(io.BytesIO(model_bytes), map_location=device)
+            model_dict = torch.load(io.BytesIO(model_bytes), map_location=device, weights_only=False)
```
Comment (mnabian) [inline physicsnemo/models/module.py]: Please explain why these changes are needed here. @CharlelieLrt for viz

File: examples/weather/flood_modeling/flood_forecaster/README.md [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+- `M40_WD_{t}.txt`: Water depth (N × 1)
+- `M40_VX_{t}.txt`: X-velocity component (N × 1)
+- `M40_VY_{t}.txt`: Y-velocity component (N × 1)
+
+### Boundary Condition Files (per timestep `t`)
+- `M40_US_InF_{t}.txt`: Upstream inflow hydrograph
+
+Data should be organized in folders with consistent naming patterns as specified in the configuration. The model supports both source and target domain datasets for domain adaptation training.
+
+**Dataset Link**: The dataset used in the paper is available at: https://github.com/MehdiTaghizadehUVa/FloodForecaster
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/README.md]: Is this correct? Are you providing the dataset, or just the scripts to generate the dataset?

File: examples/weather/flood_modeling/flood_forecaster/README.md [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+   - Adjust model parameters (channels, FNO modes, hidden dimensions)
+   - Configure training epochs, learning rates, and batch sizes
+   - Set domain adaptation hyperparameters (lambda_max, classifier architecture)
+
+### Training
+
+Run the training script:
+
+```bash
+python train.py --config-path conf --config-name config
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/README.md]: Since you already have these as the default, is it necessary to specify `--config-path` and `--config-name` here?

File: examples/weather/flood_modeling/flood_forecaster/README.md [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+1. Configure your inference settings in `conf/config.yaml`:
+   - Set `rollout_data.root` to your test dataset path
+   - Configure `rollout.out_dir` for output directory
+   - Adjust `rollout_length` and `skip_before_timestep` as needed
+
+2. Run the inference script:
+
+```bash
+python inference.py --config-path conf --config-name config
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/README.md]: Same comment here

File: examples/weather/flood_modeling/flood_forecaster/README.md [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    VY: "M40_VY_{}.txt"
+  boundary_patterns:
+    inflow: "M40_US_InF_{}.txt"
+```
+
+### Model Settings
+
+```yaml
+model:
+  model_arch: 'gino'
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/README.md]: Is this actually configurable?

File: examples/weather/flood_modeling/flood_forecaster/README.md (hunk 3 above)
Comment (mnabian) [reply examples/weather/flood_modeling/flood_forecaster/README.md]: Does your training recipe support multi-GPU/multi-node training? If yes, please specify the command. If not, please add a note.

File: examples/weather/flood_modeling/flood_forecaster/README.md (hunk 4 above)
Comment (mnabian) [reply examples/weather/flood_modeling/flood_forecaster/README.md]: Does your inference recipe support multi-GPU/multi-node inference? If yes, please specify the command. If not, please add a note.

File: examples/weather/flood_modeling/flood_forecaster/README.md [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+  doi       = {10.1016/j.jhydrol.2025.134512},
+  url       = {https://doi.org/10.1016/j.jhydrol.2025.134512}
+}
+```
+
+## Contact
+
+For questions, feedback, or collaborations:
+
+- **Mehdi Taghizadeh** – <jrj6wm@virginia.edu>
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/README.md]: Please mention that you are the code contributor and maintainer.

File: examples/weather/flood_modeling/flood_forecaster/README.md [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+- **Zanko Zandsalimi** – <mfx2uq@virginia.edu>
+- **Mohammad Amin Nabian** – <mnabian@nvidia.com>
+- **Jonathan L. Goodall** – <jlg7h@virginia.edu>
+- **Negin Alemazkoor** (Corresponding Author) – <na7fp@virginia.edu>
+
+## License
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/README.md]: You don't need this section. Please remove it.

File: examples/weather/flood_modeling/flood_forecaster/README.md [hunk 8]
```diff
@@ -0,0 +1,338 @@
+# FloodForecaster: A Domain-Adaptive Geometry-Informed Neural Operator Framework for Rapid Flood Forecasting
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/README.md]: Thanks for the comprehensive and very well structured readme!

File: examples/weather/flood_modeling/flood_forecaster/conf/config.yaml [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+  job:
+    chdir: True  # Change directory to the job's working directory.
+  run:
+    dir: ./outputs/  # Directory to save outputs.
+
+# Distributed computing
+distributed:
+  use_distributed: false
+  wireup_info: 'mpi'
+  wireup_store: 'tcp'
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/conf/config.yaml]: Would you be able to use physicsnemo distributed utilities instead?

File: physicsnemo/models/module.py (hunk 1 above)
Comment (CharlelieLrt) [reply physicsnemo/models/module.py]: That's BC breaking, we can't accept this change

**Review by CharlelieLrt — CHANGES_REQUESTED** [review]

File: examples/weather/flood_modeling/flood_forecaster/conf/config.yaml [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+  model_parallel_size: 1
+  seed: 123
+  device: 'cuda:0'
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/conf/config.yaml]: If we set `use_distributed=True`, how should the `device` filed be specified? Would it be a list of devices? Please clarify. It would be nice to add some details about distributed computing configs in the readme.

File: examples/weather/flood_modeling/flood_forecaster/conf/config.yaml [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+  fno_rank: 0.4
+  fno_fixed_rank_modes: false
+  fno_implementation: 'factorized'
+
+# Checkpoint configuration
+checkpoint:
+  save_dir: "./checkpoints_flood_forecaster"
+  resume_from_dir: null
+  resume_from_source: null
+  resume_from_adapt: null
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/conf/config.yaml]: Are these needed? If som please describe how these should be used.

File: examples/weather/flood_modeling/flood_forecaster/conf/config.yaml [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+  out_dir: "./rollout_outputs"
+
+# Optimization settings
+training:
+  n_epochs: 4
+  n_epochs_source: 2
+  n_epochs_adapt: 2
+  learning_rate: 1e-4
+  adapt_learning_rate: 1e-4
+  training_loss: 'l2'
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/conf/config.yaml]: List the available options in a comment

File: examples/weather/flood_modeling/flood_forecaster/conf/config.yaml [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+  testing_loss: 'l2'
+  weight_decay: 1e-4
+  amp_autocast: false
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/conf/config.yaml]: List the available options in a comment

File: examples/weather/flood_modeling/flood_forecaster/data_processing/data_processor.py [hunk 14]
```diff
@@ -0,0 +1,766 @@
...
+        
+        if gino is not None:
+            try:
+                return getattr(gino, name)
+            except AttributeError:
+                pass
+        
+        raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
+    
+    def save_checkpoint(self, save_folder: str, save_name: str) -> None:
```
Comment (mnabian) [inline examples/weather/flood_modeling/flood_forecaster/data_processing/data_processor.py]: I'm curious, how is GiNO's save and load checkpoint methods different from torch's save/load utilities? Couldn't you use the physicsnemo save/load utilities for better integration?

File: physicsnemo/models/module.py (hunk 1 above)
Comment (MehdiTaghizadehUVa) [reply physicsnemo/models/module.py]: Removed

File: examples/weather/flood_modeling/flood_forecaster/README.md (hunk 2 above)
Comment (MehdiTaghizadehUVa) [reply examples/weather/flood_modeling/flood_forecaster/README.md]: Updated. Repository provides data generation scripts, not the dataset.

**Comment by MehdiTaghizadehUVa** [comment]
Hi @mnabian,  All review comments are addressed:

- Deleted untrained_checkpoint.mdlus
- Removed weights_only=False from module.py
- Updated README: clarified dataset link, removed redundant flags, added multi-GPU training notes, noted single-device inference, updated contact section, removed License section
- Updated config.yaml: added comments for device (ignored in distributed mode), resume_from_source/resume_from_adapt (checkpoint paths), and listed options for training_loss/testing_loss ('l1', 'l2')
- Migrated to PhysicsNeMo checkpointing in data_processor.py

All changes are committed and ready for review.
